pkg/querier/parquet_queryable.go

```diff
@@ -0,0 +1,346 @@
+package querier
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus-community/parquet-common/schema"
+	"github.com/prometheus-community/parquet-common/search"
+	parquet_storage "github.com/prometheus-community/parquet-common/storage"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	"github.com/prometheus/prometheus/util/annotations"
+	"github.com/thanos-io/thanos/pkg/strutil"
+
+	"github.com/cortexproject/cortex/pkg/storage/bucket"
+	cortex_tsdb "github.com/cortexproject/cortex/pkg/storage/tsdb"
+	"github.com/cortexproject/cortex/pkg/storage/tsdb/bucketindex"
+	"github.com/cortexproject/cortex/pkg/tenant"
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/cortexproject/cortex/pkg/util/multierror"
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+type parquetQueryableFallbackMetrics struct {
+	blocksQueriedTotal *prometheus.CounterVec
+}
+
+func newParquetQueryableFallbackMetrics(reg prometheus.Registerer) *parquetQueryableFallbackMetrics {
+	return &parquetQueryableFallbackMetrics{
+		blocksQueriedTotal: promauto.With(reg).NewCounterVec(prometheus.CounterOpts{
+			Name: "cortex_parquet_queryable_blocks_queried_total",
+			Help: "Total number of blocks found to query.",
+		}, []string{"type"}),
+	}
+}
+
+type parquetQueryableWithFallback struct {
+	services.Service
+
+	queryStoreAfter       time.Duration
+	parquetQueryable      storage.Queryable
+	blockStorageQueryable *BlocksStoreQueryable
+
+	finder BlocksFinder
+
+	// Subservices manager.
+	subservices        *services.Manager
+	subservicesWatcher *services.FailureWatcher
+
+	// metrics
+	metrics *parquetQueryableFallbackMetrics
+}
+
+func NewParquetQueryable(
+	config Config,
+	storageCfg cortex_tsdb.BlocksStorageConfig,
+	limits BlocksStoreLimits,
+	blockStorageQueryable *BlocksStoreQueryable,
+	logger log.Logger,
+	reg prometheus.Registerer,
+) (storage.Queryable, error) {
+	bucketClient, err := bucket.NewClient(context.Background(), storageCfg.Bucket, nil, "parquet-querier", logger, reg)
+
+	if err != nil {
+		return nil, err
+	}
+	manager, err := services.NewManager(blockStorageQueryable)
+	if err != nil {
+		return nil, err
+	}
+
+	cDecoder := schema.NewPrometheusParquetChunksDecoder(chunkenc.NewPool())
+
+	parquetQueryable, err := search.NewParquetQueryable(cDecoder, func(ctx context.Context, mint, maxt int64) ([]*parquet_storage.ParquetShard, error) {
+		userID, err := tenant.TenantID(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		blocks, ok := ExtractBlocksFromContext(ctx)
+		if !ok {
+			return nil, errors.Errorf("failed to extract blocks from context")
```

Comment:
I wonder why it is consider an error. If there is no block found, we can still find the key? So we consider no key as error?

Reply:
We should never call this without setting this key.. If we do so.. it should be an error.
If no parquet blocks are found we should just bypass this queryable or set an empty slice to make sure we did not accidentally call this queryable without setting the blocks before.